src/hotspot/share/code/codeCache.cpp

```diff
@@ -324,34 +307,44 @@ void CodeCache::initialize_heaps() {
     }
   }
 
+  // last adjustment: leftovers from page alignment go to non_nmethod segment
+  non_nmethod.size += non_profiled.size & alignment_mask(min_size);
+  non_nmethod.size += profiled.size & alignment_mask(min_size);
```

Comment:
Why do we need this code?

Reply:
Let us assume that with the previous calculations we have this layout:
- {cache_size=18004KB, non_profiled.size = 6002KB, profiled.size = 6002KB, non_nmethod.size=6000KB}

With next align_down for profiled and non_profiled segments, we get:
- {cache_size=18004KB, non_profiled.size = 6000KB, profiled.size = 6000KB, non_nmethod.size=6000KB}

My intention here is to put 4KB into the non_nmethod segment to make the sum of the segment sizes equal to cache_size:
- {cache_size=18004KB, non_profiled.size = 6000KB, profiled.size = 6000KB, non_nmethod.size=6004KB}

Reply:
Why has `non_nmethod` been chosen? I have not seen it running out of space. It usually has a lot of free space. Also in case of large pages, e.g. 2M, it can get even more space than needed.
I'd rather give additional space to `profiled` or `non_profiled`. As `profiled` is optional, to simplify the code we can have:
```
  non_nmethod.size = align_up(non_nmethod.size, min_size);
  non_profiled.size = align_down(non_profiled.size, min_size);
  profiled.size = align_down(cache_size - non_profiled.size - non_nmethod.size, min_size);
```

For you example, we will have:

{cache_size=18004KB, non_profiled.size = 6004KB, profiled.size = 6000KB, non_nmethod.size=6000KB}

Reply:
My reasons are: the logic was before the change and we know non_nmethod segment always exist. Although, ok, let it be non_profiled segment to get an extra load.